File: SourceAFIS/Extraction/Model/PoreRemover.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Drawing;
using SourceAFIS.General;
using SourceAFIS.Meta;

namespace SourceAFIS.Extraction.Model
{
    public sealed class PoreRemover : ISkeletonFilter
    {
        [DpiAdjusted]
        [Parameter(Lower = 3, Upper = 100)]
        public int MaxArmLength = 45;

        [Nested]
        public KnotRemover KnotRemover = new KnotRemover();

        public DetailLogger.Hook Logger = DetailLogger.Null;

        public void Filter(SkeletonBuilder skeleton)
        {
            foreach (SkeletonBuilder.Minutia minutia in skeleton.Minutiae)
            {
                if (minutia.Ridges.Count == 3)
                {
                    for (int exit = 0; exit < 3; ++exit)
                    {
                        SkeletonBuilder.Ridge exitRidge = minutia.Ridges[exit];
                        SkeletonBuilder.Ridge arm1 = minutia.Ridges[(exit + 1) % 3];
                        SkeletonBuilder.Ridge arm2 = minutia.Ridges[(exit + 2) % 3];
                        if (arm1.End == arm2.End && exitRidge.End != arm1.End && arm1.End != minutia && exitRidge.End != minutia)
                        {
                            SkeletonBuilder.Minutia end = arm1.End;
                            if (end.Ridges.Count == 3 && arm1.Points.Count <= MaxArmLength && arm2.Points.Count <= MaxArmLength)
                            {
                                arm1.Detach();
                                arm2.Detach();
                                SkeletonBuilder.Ridge merged = new SkeletonBuilder.Ridge();
                                merged.Start = minutia;
                                merged.End = end;
                                foreach (Point point in Calc.ConstructLine(minutia.Position, end.Position))
                                    merged.Points.Add(point);
                            }
                            break;
                        }
                    }
                }
            }
            KnotRemover.Filter(skeleton);
            Logger.Log(skeleton);
        }
    }
}

File: java/sourceafis/src/sourceafis/extraction/model/PoreRemover.java
/**
 * @author Veaceslav Dubenco
 * @since 20.10.2012
 */
package sourceafis.extraction.model;

import sourceafis.general.Calc;
import sourceafis.general.DetailLogger;
import sourceafis.general.Point;
import sourceafis.meta.DpiAdjusted;
import sourceafis.meta.Nested;
import sourceafis.meta.Parameter;

/**
 * 
 */
public class PoreRemover implements ISkeletonFilter {

	@DpiAdjusted
	@Parameter(lower = 3, upper = 100)
	public int MaxArmLength = 41;

	@Nested
	public KnotRemover KnotRemover = new KnotRemover();

	public DetailLogger.Hook Logger = DetailLogger.off;

	@Override
	public void Filter(SkeletonBuilder skeleton) {
		for (SkeletonBuilderMinutia minutia : skeleton.getMinutiae()) {
			if (minutia.getRidges().size() == 3) {
				for (int exit = 0; exit < 3; ++exit) {
					Ridge exitRidge = minutia.getRidges().get(exit);
					Ridge arm1 = minutia.getRidges().get((exit + 1) % 3);
					Ridge arm2 = minutia.getRidges().get((exit + 2) % 3);
					if (arm1.getEnd() == arm2.getEnd()
							&& exitRidge.getEnd() != arm1.getEnd()
							&& arm1.getEnd() != minutia
							&& exitRidge.getEnd() != minutia) {
						SkeletonBuilderMinutia end = arm1.getEnd();
						if (end.getRidges().size() == 3
								&& arm1.getPoints().size() <= MaxArmLength
								&& arm2.getPoints().size() <= MaxArmLength) {
							arm1.Detach();
							arm2.Detach();
							Ridge merged = new Ridge();
							merged.setStart(minutia);
							merged.setEnd(end);
							for (Point point : Calc.ConstructLine(
									minutia.getPosition(), end.getPosition()))
								merged.Points.add(point);
						}
						break;
					}
				}
			}
		}
		KnotRemover.Filter(skeleton);
		Logger.log(skeleton);
	}
}
